hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterClientProtocol.java

```diff
@@ -734,6 +811,13 @@ public boolean mkdirs(String src, FsPermission masked, boolean createParent)
         new Class<?>[] {String.class, FsPermission.class, boolean.class},
         new RemoteParam(), masked, createParent);
 
+    // Auto-create user home dir for a trash path.
+    // moveToTrash() will first call fs.mkdirs() to create the parent dir, before calling rename()
+    // to move the file into it. As a result, we need to create user home dir in mkdirs().
+    if (autoCreateUserHomeForTrash) {
+      createUserHomeForTrashPath(src, locations);
```

Comment:
I don't think you should be piggybacking on {{mkdirs}} to get this created. If this is a useful feature to have, I'd suggest adding a new FileSystem API named something like {{provisionTrashPath()}} similar to {{provisionEZTrash}} and call it in {{TrashPolicy#moveToTrash}} before {{mkdirs}}. 

Can you also call out why you need this vs. provisioning user trash paths as part of the (off band) process of provisioning the user home directories on HDFS?

Reply:
Given that we already have provisionEZTrash(), it does make sense to do the same for this.